### PR TITLE
considerate bus stop behaviour

### LINKED ISSUE #1688: Improve bus stop path
image
buses follow the red path in the image above. and tend to block the road while stopping particularly when there are multiple buses.
it would be better if they followed the black path to get out of the way as fast as possible.
if Articulated bus turns too sharp it may block the inner lane so the path should not be too sharp.
image

related: https://github.com/CitiesSkylinesMods/TMPE/issues/1020#issuecomment-716743962

### PR DESCRIPTION
**Pull request by kianzarrin**
fixes #1688. see issue for details
[TMPE.zip](https://ci.appveyor.com/api/projects/krzychu124/tmpe/artifacts/TMPE.zip?branch=1688-considerate-buses)

### CODE REVIEW

File: TLM/TLM/Patch/_NetLane/CalculateStopPositionAndDirection.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+namespace TrafficManager.Patch._NetLane; 
+using ColossalFramework.Math;
+using HarmonyLib;
+using System;
+using TrafficManager.Util;
+using UnityEngine;
+
+[HarmonyPatch(typeof(NetLane), nameof(NetLane.CalculateStopPositionAndDirection))]
```
[the remark this answers is not in the document]
Comment (kianzarrin) [reply TLM/TLM/Patch/_NetLane/CalculateStopPositionAndDirection.cs:8]: good point.

[the remark this answers is not in the document]
Comment (kianzarrin) [reply TLM/TLM/Patch/_NetLane/CalculateStopPositionAndDirection.cs:8]: they all work

**Review by krzychu124 — APPROVED** [review]
I've tested with bus, trolleybus, and trams (obviously shouldn't change anything, but just in case), everything seems to be working nicely. Noticeably better than vanilla. Performance wise most likely in margin of error territory which is nice. 

One thing I noticed, but I haven't had time to compare it to scenario without the change is that when you put bus stop at segment joined directly to big intersection node busses going from other segment may overshoot stop position and run into each other instead of forming a waiting queue especially if they are turning from different direction. 

_ugly example when approaching green bus is not able to stop without hitting blue bus that is already waiting at the bus stop_
image

**Comment by kianzarrin** [comment]
@krzychu124 if you have a test city I'd like to take a look. I'd be surprised if my changes do anything to contribute to this situation since I am only changing the sideways position by a little bit.

**Comment by krzychu124** [comment]
> @krzychu124 if you have a test city I'd like to take a look. I'd be surprised if my changes do anything to contribute to this situation since I am only changing the sideways position by a little bit.

Yeah, but maybe changes to path are increasing a chance to collision detection issues e.g.: because target positions are generated differently further exposing its poor accuracy at high speed

**Comment by kianzarrin** [comment]
@krzychu124 do you have saved city? it might be hard to reproduce

**Comment by krzychu124** [comment]
> @krzychu124 do you have saved city? it might be hard to reproduce

Yeah, I have, but I don't see any difference between this build and current `master` (on both sometimes they do weird things) so I think we can ignore that for now.

**Review by kvakvs — APPROVED** [review]

**Comment by kianzarrin** [comment]
https://github.com/CitiesSkylinesMods/TMPE/issues/1695

**Pull request merged**